backend/business-partner-agent/src/main/java/org/hyperledger/bpa/impl/ChatMessageEventHandler.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright (c) 2020-2021 - for information on the respective copyright owner
+ * see the NOTICE file and/or the repository at
+ * https://github.com/hyperledger-labs/business-partner-agent
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.hyperledger.bpa.impl;
+
+import lombok.NoArgsConstructor;
+import lombok.NonNull;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.StringUtils;
+import org.hyperledger.aries.api.message.BasicMessage;
+import org.hyperledger.bpa.impl.aries.ConnectionManager;
+import org.hyperledger.bpa.model.ChatMessage;
+import org.hyperledger.bpa.repository.PartnerRepository;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import java.util.UUID;
+
+@Slf4j
+@NoArgsConstructor
+@Singleton
+public class ChatMessageEventHandler {
+
+    @Inject
+    PartnerRepository partnerRepo;
+
+    @Inject
+    ConnectionManager cm;
+
+    @Inject
+    ChatMessageManager chatMessageManager;
+
+    @Inject
+    NotificationManager notificationManager;
+
+    public void handleIncomingMessage(@NonNull BasicMessage basicMessage) {
```

Comment:
I think both of these methods belong more to a manager as they describe flows. To me a handler is basically just a thin layer that either handles single step tasks directly (database state update) or does a preselection based on event state and then dispatches the event to a manager to do the heavy lifting. Also I would like to keep the separation between webhook (handler) and user triggered events (manager). Basic Massages are a bad example for a handler as they only have a single state and could also be handles by a manager directly.

Reply:
I think this area of logic causes a lot of confusion. We had discussed what to do with a workflow where the logic is done by many "managers" or services. Where does the separation of concerns (ie. create a task, notify frontend etc) get coordinated? If it is in one manager (ie. ConnectionManager) then it needs to coordinate to the next Manager (ie ActivityManager). I was doing that via application side events, but that wasn't liked because it is difficult to follow or see all the things that happen when a particular "event" (either Aries event or user frontend click). So I thought we had come to the conclusion that we wanted an intermediate layer that would act as the coordinator - some event happens (either Aries or user initiated), and one class would control the flow of work and coordination. What I think you are saying is Aries events are coordinated one way (via a specific event handler class - connectioneventhandler, proofeventhandler) and user initiated are done via Manager.  Where the confusion lies is that the eventhandler coordinates all the required Managers to do the Aries related work, then a Manager has to coordinate other Managers to do all the work for user initiated flows. So a Manager is both coordinated externally AND then acts as a coordinator.  

I think consistency is important, so we should decide what we are doing... even in the simplest case (like basic messaging), it should still follow the pattern we want to go with; even though it COULD be done in a single class. I am getting confused on what we are doing. We don't want to use application-side events between Managers because it scatters and duplicates logic (like fetching partners repeatedly), but we don't want to have Managers be the coordinator between Managers on Aries events, but we do on user events? So, are we breaking down the Managers to have more public methods (smaller scope) so those methods are used in the *EventHandler (coordinating calls from many Managers), AND then the Manager being the coordinator/workflow between other Managers (again using the smaller methods) for incoming/user functions?  Perhaps we need to draw up a Naming and Scope of concerns convention.